### PR TITLE
COP-10048: Add TextArea component

### PR DESCRIPTION
**Pull request by paulfgraham**
### Description
Added a `TextArea` component (initially as part of a demo). The Storybook stories are aligned with those on GDS (with adjustments and additions where appropriate) and the unit tests are quite extensive and sensible (100% coverage, catching all the edge cases I could think of).

https://support.cop.homeoffice.gov.uk/browse/COP-10048

### To test
Essentially covered by accompanying unit tests and Storybook stories.

### CODE REVIEW

File: src/assets/scss/storybook.css [hunk 1]
```diff
@@ -108,3 +108,9 @@ div[role="main"] > div {
 #docs-root .sbdocs.sbdocs-wrapper {
   padding-top: 1.75rem;
 }
+
+#docs-root .sbdocs-p {
```
Comment (paulfgraham) [inline src/assets/scss/storybook.css:112]: This chunk of CSS is to align the Storybook appearance with the GDS appearance - everything was already pretty close and it was just paragraphs that had text that was too small.

File: src/TextArea/TextArea.jsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  error,
+  readonly,
+  classBlock,
+  classModifiers: _classModifiers,
+  className,
+  ...attrs
+}) => {
+  const [rows, setRows] = useState(DEFAULT_ROWS);
+  useEffect(() => {
+    const rowsNumber = parseInt(_rows, 10);
```
Comment (gardnersam) [inline src/TextArea/TextArea.jsx:26]: What's the benefit in letting rows be anything other than an int or PropTypes.number?

File: src/TextArea/TextArea.jsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  readonly: PropTypes.bool,
+  classBlock: PropTypes.string,
+  classModifiers: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
+  className: PropTypes.string
+};
+
+TextArea.defaultProps = {
+  rows: DEFAULT_ROWS,
+  classBlock: DEFAULT_CLASS,
+  classModifiers: []
```
Comment (gardnersam) [inline src/TextArea/TextArea.jsx:65]: Is this needed? It gets set on line 29 anyway.

File: src/TextArea/TextArea.jsx [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  readonly,
+  classBlock,
+  classModifiers: _classModifiers,
+  className,
+  ...attrs
+}) => {
+  const [rows, setRows] = useState(DEFAULT_ROWS);
+  useEffect(() => {
+    const rowsNumber = parseInt(_rows, 10);
+    setRows(isNaN(rowsNumber) ? DEFAULT_ROWS : rowsNumber);
```
Comment (gardnersam) [inline src/TextArea/TextArea.jsx:27]: Will this cause a re-render to set rows to DEFAULT_ROWS when it's already set to that? Not sure if that really matters to be fair...

File: src/TextArea/TextArea.stories.mdx [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+Use the textarea component when you need to let users enter an amount of text that's longer than a single line.
+
+# When not to use this component
+Users can find open-ended questions difficult to answer.
+It might be better to break up one complex question into a series of simple ones,
+for example where users can select from options using <Link href="./?path=/docs/d-radios--radios">radios</Link>.
+
+### If you need to ask an open question
+Do not use the textarea component if you need to let users enter shorter answers no longer than a single line,
+such as a phone number or name. In this case, you should use the <Link href="./?path=/docs/d-textinput--default-story">text input</Link>
```
Comment (gardnersam) [inline src/TextArea/TextArea.stories.mdx:43]: Other stories use the [link text]: https://actual-link.com format for this. This is possibly nicer but do we want a consistent approach?

File: src/TextArea/TextArea.test.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
+import React from 'react';
```
Comment (gardnersam) [inline src/TextArea/TextArea.test.js:1]: I don't think we need this, I've not been including it myself. Do we want to make the standard that it's in anyway and I'll start following that?

File: src/TextArea/TextArea.jsx (hunk 3 above)
Comment (paulfgraham) [reply src/TextArea/TextArea.jsx:65]: Yes, because the property arrives in the component as `_classModifiers` so we default it to an empty array, which means we can _use_ it on line 29. We have the same in place for the other components.

File: src/TextArea/TextArea.jsx (hunk 4 above)
Comment (paulfgraham) [reply src/TextArea/TextArea.jsx:27]: It _shouldn't_ do, as it doesn't represent a change to the DOM. Internally, I think React will "re-render", but when it compares what it has to the DOM and finds no changes, it will do nothing.

If I check the value I'm _going_ to set `rows` to against the existing value of `rows`, I need to include it in the `useEffect` array, which would mean it would fire a second time whenever we _do_ want to update it. Swings and roundabouts on that one, I think. 🤔

File: src/TextArea/TextArea.jsx (hunk 2 above)
Comment (paulfgraham) [reply src/TextArea/TextArea.jsx:26]: It's a usage thing. I mentioned this on the commit message but not as a comment in here.

Basically, when it's passed from the Form Renderer, it'll be a number... but if someone simply wants to use a `TextArea` from the component library directly, I would have thought they'd want to simply do this:
```
<TextArea id="details" fieldId="details" rows="8" />
```

If we only allow numbers, it'll require people to do this, which I think is less intuitive:
```
<TextArea id="details" fieldId="details" rows={8} />
```

File: src/TextArea/TextArea.stories.mdx (hunk 5 above)
Comment (paulfgraham) [reply src/TextArea/TextArea.stories.mdx:43]: I think the other stories _might_ need to use the `<Link />` syntax simply because of the GovUK styling - the `[link text](actual link)` syntax uses the Storybook styling.

The alternative is that we change the Storybook styling to match that of `govuk-link`.

File: src/TextArea/TextArea.test.js (hunk 6 above)
Comment (paulfgraham) [reply src/TextArea/TextArea.test.js:1]: Whenever I have something like line 18, I prefer to import `React`. I don't think `yarn lint` will complain if it's not there... but it's also not telling me it's an unused import so it seems like a preference thing. 🤷‍♂️

Comment (paulfgraham) [reply src/TextArea/TextArea.test.js:1]: Also for this, when I've looked at documentation around jest testing within React, they always seem to import `React` even when they don't appear to be doing anything different to what I'm doing in my tests... so maybe it's standard practice? Or maybe _not_ including it is standard practice. Not sure. 😂

**Review by gardnersam — APPROVED** [review]
Happy with all comment responses, looks good to me!

File: src/TextArea/TextArea.test.js (hunk 6 above)
Comment (paulfgraham) [reply src/TextArea/TextArea.test.js:1]: As discussed offline with @gardnersam:
https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html

It seems there's no need to do `import React from 'react';` in React v17 onwards... but we're using v16 so we ought to keep it in place. It's also perfectly _okay_ to do it in v17, just not _necessary_. If this library is ever upgraded beyond v16, a decision on best practice can be made then. 😁

**Review by philjlharvey — APPROVED** [review]
Approved on behalf of Ali Khattab and Keenen Bratley, who are happy with all the changes.
